### ISSUE TITLE
Add a nice GUI to improve user experience

### ISSUE THREAD

**Issue by titan73** [wine-compholio/wine-staging#263]
This needs to be cross platformand should for with Linux, FreeBSD and Mac OSX.
If possible using a language and toolkit that are standard and have minimum dependencies with other packages.
This does not require any Windows knowledge but only how Wine works.

**Comment by slackner** [comment]
I've been thinking about that since some time, and I'm still not sure yet what the best way is to write such a GUI.

First of all, we should probably keep such a GUI separate from our main package - some distros don't like `winetricks` and similar apps. I would also like to avoid dependencies on bigger toolkits like GTK/QT - this is not really a final decision, but I hate it when packages have a lot of external dependencies for no good reason.

Some features I would definitely like to have:

* Management of wineprefixes. Either using some configuration files, or just recognizing everything like ~/.wine*/system.reg as a wineprefix.

* Similar to PlayOnLinux we could offer predefined scripts for application, or at least allow to easily apply winetricks recipes on a prefix. Its a bit bad that winetricks doesn't offer any methods to undo such changes... I am wondering how difficult it would be to implement something like that ;)

* Show the applications installed in each prefix, and allow to start them from the GUI. Including builtin wine tools like `winecfg`.

We should keep in mind that this should not just be a PlayOnLinux-clone, otherwise its useless.

**Comment by titan73** [comment]
I'm thinking. What about using wine for the GUI? We could have a prefix dedicated to this GUI and avoid additinal dependencies. This may be not that nice albeit gdi+ may help.

**Comment by titan73** [comment]
Or maybe mshtml/javascript if wecan access the filesytem and span a process from there.

**Comment by DarkPlayer** [comment]
I think we should detect existing wine prefixes. My idea was to enumerate the installed software in these prefixes (either by detecting supported programs or by taking a look at the startmenu entries) and to provide an overview to the user from which he can directly start these programs. 

Many people don't want to pollute their desktop or start menu with wine programs so I think this would be a nice solution for this problem. In order to convince new users of this GUI it should work with zero configuration and directly find the installed software.

I created a quick mockup how the main windows might look like from my point of view:

![wine-gui-mockup](https://cloud.githubusercontent.com/assets/3102472/5696893/a7e068c2-99e0-11e4-870f-aa227ef31e34.png)

The overview should basically behave like the windows desktop from which you can start programs and it might also be a good idea to add a search function (simply by entering the search string while the program is focused). In case you want to improve this mockup, you can find the svg file [here](https://dl.dropboxusercontent.com/u/61413222/wine-gui-mockup.svg).

@titan73 Although using Wine for displaying the GUI might solve the dependency problem, I am not sure if this is a good idea. Wine takes quite some time to boot and you might want the GUI to come up quickly. Moreover using a scripting language like python would make some things easier.

**Comment by DarkPlayer** [comment]
I don't know if this is a good idea but I still want to hear your opinion on this @titan73 @slackner : We could run a http server in python which serves the GUI as website. If the user starts our program, the script would start a HTTP server binded to localhost and then open the web browser.

This would allow us to completely get rid of the GUI framework dependency and we could create a fancy GUI using CSS effects etc. Thanks to AJAX and web sockets the application could be as interactive as a normal program.

The problem of this solution is that it does not provide a very good security. Every program which can send http packages to localhost may start a wine process. The issue would also appear for programs that  can send packages to the X server if we use a normal GUI toolkit, but the http idea is easier to exploit. We can try to reduce the risk by passing a secret token to the web browser but there is no perfect solution to this problem.

So what do you guys think about this idea?

**Comment by slackner** [comment]
@DarkPlayer: It also sounds like an interesting idea, but I am concerned a bit about the security. Of course I know that basically any other server (including the wineserver itself) also suffers from security issues, but it is less likely that such a connection is exploited to do something harmful. The only workaround for that is probably to use an embedded webbrowser and security tokens.... but not sure :/ I somehow would still prefer a standalone solution, which is also often a lot faster.

Advantage of using a website - it is relatively easy to integrate external resources. For example a website like Appdb could offer direct links to quick-install various programs in wineprefixes. Needs proper security checks though, and probably also user confirmation...

**Comment by damianatorrpm** [comment]
@slackner @DarkPlayer: I'd love to contribute to your wonderful project and have thought about this myself recently, but I'm only confident in Qt, so if Qt is ok...

**Comment by slackner** [comment]
I am also fine with using Qt. Since we most probably will have to provide it as a separate package (instead of directly including in wine-staging) the additional dependencies do not really matter that much.

Not sure if you saw it, but we also discussed this topic on IRC a bit, and what we should also keep in mind is that starting from scratch is always a very big task - especially writing scripts for installing various games/apps takes a lot of time. We should probably think about partially reusing work from existing projects (like POL/...) or maybe even start some cooperation to avoid doing all that work again. Besides prefix management this is one of the most common reasons why people prefer to use a GUI.

@damianatorrpm: You're welcome to give it a try of course, but you should probably show us your initial idea (more detailed descriptions or even better a first prototype) before you spend too much time working on it, so that we can still include improvements. We also have no good solution yet how to create/manage thousand of install scripts, which is something we should try to solve before focusing too much on a specific solution.

**Comment by damianatorrpm** [comment]
@slackner: I had quite some ideas on that. 
Actually I would include event the functionality of winecfg into that and extend it (e.g proxy options and stuff) - all shiny and user friendly.

One idea I have it to automate the process and eliminating the need for install new app => choose app at all and collect a database with hashes of programs and/or other characteristics and matching them automatically by that (since this will be choosen at install updates should be no problem since winetricks etc already done). 
The idea here is that any *.exe is opened by
$ourWineWrapper << check_for_known_hash <<process_to_wine

The install scripts should be share with other projects of course, I can get in touch with them.
Another idea is an addition to the mentioned above:
$ourWineWrapper << check_for_known_hash << choose prefix to install to(only if installers) << process_to_wine

So actually until here only the checkbox to choose the prefix.
The gui can be started to manage prefixes and it's options.

The GUI would have the list of the prefixes and also button configure prefix, that will have winecfg like functionality.

**Comment by titan73** [comment]
The http server solution is a bit heavier that Qt but I have no particular preference. I think Qt (or Gtk) is a simpler solution to start with and it's better to have a solution people would like to tackle.

Regarding the GUI itself, I've never used CrossOver nor POL GUIs so I couldn't tell what could be the right GUI. It's clear CrossOver one is not suitable. POL GUI is probably closer to what's we would like.
I think it's important to mention that compared to POL, this GUI is meant to be used with non hacked versions of wine and should enforce people to use the lastest version of wine stating. I think this GUI should not only be used to make use life easier but also to get feedback from user and help giving information, debug log, ... to wine-staging devs. We were talking about all these bug reports without any useful information. This GUI may help the process to report bugs to the right place and help generating logs for devs.

**Comment by titan73** [comment]
Can someone tells how is POL GUI made? Which language and toolkir?
We can start our own GUI or fork POL to add our new features and reuse POL scripts first. In parallel we can discuss with POL guys if they can base more games on wine-staging and how we can help the process to give feedback to devs.
It would be good to define the key principles and then start with a first implementation and see how think goes and how it can be improved.

**Comment by damianatorrpm** [comment]
@titan73: It's python and wx_widget, nothing I have an idea of.

**Comment by DarkPlayer** [comment]
If you don't choose the browser solution, there are basically 3 GUI toolkits that might be interesting for us.

- **wxWidgets**: The aim of wxWidgets is to use the controls provided by the system, for example the common controls on Windows. This should make the application look more native compared to the other toolkits which completely draw the controls on their own and look the same on all platforms. The downside of this approach is that wxWidgets is not so good in abstracting the controls and you sometimes run into platform specific issues (this is at least my experience from the last time I tried it out). Using wxWidgets would increase the time we need to spend on testing if we want to support MacOS X in the future. Moreover wxWidgets is rarely used on Linux (maybe because Linux does not have native controls ;-)) and would pull in a new dependency on most systems.

- **GTK**: GTK is already installed on a lot of Linux systems and provides everything you need for an advanced GUI. There are also bindings for most programming languages. I am not so happy about the documentation of the python bindings, but they are still usable.

- **Qt**: Qt is basically like GTK with the difference that it does not only provide a GUI toolkit but also contains other stuff like networking functions ([list of modules](http://doc.qt.io/qt-5/qtmodules.html)).

I think the decision for the GUI Toolkit also depends on the programming language we use. If you write the program in Python, you don't necessarily need the networking functions of Qt but you can directly use the ones provided by Python, so GTK might be a more lightweight choice. If you want to write the program in C/C++ you might be happy about the HTTP library of Qt.

**Comment by LoathingKernel** [comment]
Why nobody has suggested upon improving Q4Wine? Is there an issue that I am not seeing, except for the dependency on Qt.

It has most of the things described as needed in this issue. I has prefix management, it can clean-up wine generated menu entries, it has winetricks integration. So unless there is some underlying issue that I fail to see, it seems like a really good starting point.

**Comment by slackner** [comment]
I have to admit that I didn't really look much at existing projects yet. I just tested it a bit, and one of the first things I immediately noticed is that it is not really user-friendly.

Unexperienced users will have no clue what a lot of these settings mean (10 tabs full of settings when creating a new prefix), and I guess even experienced users will never have to change everything. This is just a bit bad UI design, normally all those advanced settings should be hidden in some configuration menus. Also, I don't think Q4Wine offers automated installation of Windows software, which was also one of the ideas we discussed.

If we want to use Q4Wine as a base, I am pretty sure we would have to change a lot to make it as easy as PlayOnLinux/Crossover.

**Comment by DarkPlayer** [comment]
I don't have so much experience with Q4Wine but from what I have seen so far it looks like Q4Wine is targeting advanced users by providing almost all Wine settings through a GUI. This is not directly our aim. We would like to create a GUI for beginners.

The GUI should provide typical functions to install software in Wine and hide all advanced features by default. For example if the user wants to install steam, this should basically be a one click operation. You select the software from a list / search for it and the GUI downloads Steam, creates a new prefix and starts the installation and maybe disables dwrite in this prefix to workaround the dwrite bug. There should also be the possibility to select the Wineprefix in which you want to install Steam, but this should be hidden on an advanced tab / dropdown menu.

**Comment by compholio** [comment]
So, part of the reason I was working with "pip" the other day was to see if I could compile a Windows PySide Python app using PyInstaller.  I can confirm that this works.  I think a huge benefit of taking a route like this is that we can provide a native application for supported systems and a compiled Windows app for systems we cannot reach otherwise.

**Comment by DarkPlayer** [comment]
@compholio I don't think this is suitable solution. Configuring other wineprefixes and starting programs in other wineprefixes would become very ugly.

**Comment by compholio** [comment]
@DarkPlayer: How so? Just use a special prefix for the config tool and look in Z:/$HOME/ for configuring prefixes.  It's also relatively easy to launch a script that will start a session for the requested prefix.

**Comment by DarkPlayer** [comment]
I think we can discuss a long time about what framework would be the best, but honestly most of us are already quite busy with fixing wine bugs in their spare time. @damianatorrpm would like to contribute by writing a GUI in Qt and I think we should stick to this solution. Qt is available for Linux, Mac OS X and FreeBSD and therefore fulfills our requirements. It might be a bit heavy but wine isn't small either.

If @damianatorrpm is still interested in helping us I would suggest writing a small prototype so that we get an overview about the required tasks. When we have a working prototype we may speak about the details, for example how we want to write the installation scripts, maybe we utilize some small embedded scripting language like lua or just use simple bash scripts. 

@slackner @compholio: Are you fine with this approach?

@damianatorrpm I would suggest you to join our IRC channel #wine-staging on FreeNode if you have any questions or if you would like us to add some special feature to Wine which might simplify your development.

**Comment by slackner** [comment]
@DarkPlayer: I am fine with that idea.

**Comment by compholio** [comment]
@DarkPlayer: Yup, if @damianatorrpm happens to use PySide I'll be _really_ happy since that's one of the things I've been playing with for work recently ;)

**Comment by damianatorrpm** [comment]
@slackner @DarkPlayer:
I will prototype a GUI over the weekend.

@compholio: I've not used PySide yet, but I guess it would't be too much harder to c++, which I prefer.

**Comment by damianatorrpm** [comment]
![screenshot from 2015-01-18 16 45 16](https://cloud.githubusercontent.com/assets/6593358/5799676/dcb2aec2-9fdd-11e4-946b-73c599afba37.png)

**Comment by damianatorrpm** [comment]
![screen1](https://cloud.githubusercontent.com/assets/6593358/5799680/e8652844-9fdd-11e4-8c0c-cfb52cb4feb2.png)

**Comment by damianatorrpm** [comment]
This is what I think as wrapper to detect installers. Opinions?
How about where possible only that and pass to the installer silent parameters?

**Comment by slackner** [comment]
@damianatorrpm: Thanks for your effort! Without seeing a bit more complete GUI prototype than a single screenshot its difficult to imagine how all the pieces will fit together though. Looking forward to see the rest of your ideas ;)

Feedback for this part:
* Where can you select for example which app should be installed? Or will that be part of the "parent window"?
* Can this installer be only used for software which is available online or also for installing from CD-ROM? Depending on the installer it might be necessary to specify one or more paths to the installer files. Sometimes the user will also have to manually download software. Some software also has dependencies (for example .NET, specific native dlls, other winetricks recipes, ...), the installer should be able to handle that, and offer a nice and convenient way to install everything required.
* Instead of specifying the full installation path (which is often not possible) it would be better to just select one of the existing prefixes. Usually its sufficient to install to some "default" location. The official recommendation is to install only one program per prefix, which makes it unnecessary to use a different path.

Ideas for the main GUI:
* Similar to a file browser we could show a list of existing wine prefixes on the left side. Maybe prefixes could even have user-defined icons assigned, for example the application icon for which the prefix is used. A user could for example have multiple prefix "Game 1", "Game 2", "Game 3".
* Right clicking on the list of prefixes or special buttons could allow creating / deleting prefixes.
* After clicking on a specific prefix the right side of the GUI shows the list of installed programs and/or desktop symbols. Programs can be started by double-clicking on an icon.
* To install new software there could be for example a "Manage installed programs" button, which uses the currently selected wineprefix and opens an installer/uninstaller GUI. The same window could also be used to install various winetricks recipes, and to list which software/recipes are already installed.

**Comment by damianatorrpm** [comment]
@slackner What do you mean with:
>Where can you select for example which app should be installed? Or will that be part of the "parent window"?

I mean instead of opening .exe with wine, wine is renamed to wine-bin and wine is a little script that checks if setup or installer is detected and based on that
a) It shows the above screen if it detects that the installer belongs to a program with known issues and we need to execute stuff 
b) it shows above screen for all installers 
You can quite confidently identify installers by checking for setup.exe/installer.exe, of course this doesn't cover a 100% but for the rest you can have other rules.

This would cover IMHO the whole installation part and we could even try to hide the setup.exe itself by passing silent parameters. Where  for example .NET, specific native dlls, other winetricks recipes is required, the installer will detect that and execute it automagically.

After that the main window would be the "Manage Prefixes and it's Configurations stuff".

>Instead of specifying the full installation path (which is often not possible) it would be better to just select one of the existing prefixes. Usually its sufficient to install to some "default" location. The official recommendation is to install only one program per prefix, which makes it unnecessary to use a different path.

Yes. I thought of trying to detect the name of the program one is installing and based on that make suggestions like /home/user/wine-programs/$

I'll try to get a working prototype that works with some example program like vlc or so and share it with you soon.

**Comment by damianatorrpm** [comment]
@slackner @DarkPlayer 
opinions?

**Comment by DarkPlayer** [comment]
I am not sure whether it is a good idea to replace the original wine executable with a script. You should still provide the user with the possibility to use wine without the gui. The method of adding windows executables to binfmt seems to be deprecated and therefore it is unlikely that a user accidentally starts a program using wine. 

The most likely use case is to fire up the gui to install a s specific software from a list of supported programs. Your program would then either download the application or show a dialog to select the CD drive containing the installation disc. Installing unsupported software might be interesting, especially when the GUI does not support so much software at the beginning, but it is not necessary for the prototype.

**Comment by slackner** [comment]
I have to agree to @DarkPlayer, I also see the GUI mainly as an addition, not as a replacement for the main "wine" binary. There are also a lot of programs which try to interact with wine, and they will be confused when it doesn't behave as expected and opens a GUI instead of running the given command (winetricks, pipelight, ....). I like the idea of automatic detection of installers, but we should not "integrate" it too much, and instead keep it as a separate executable.

In addition to @DarkPlayer's comments I would also like to point out that you should not forget prefix management. This is also very important, especially for unexperienced users. Running "wine setup.exe" is relatively easy when the installer works out of the box, but its not easy to find where it was actually installed, how to start it, how to uninstall it later or start from scratch with a new virtual environment if something doesn't work, ... ;)

**Comment by damianatorrpm** [comment]
@slackner @DarkPlayer 
http://wiki.winehq.org/ThirdPartyApplications
Above an extensive list of wine-gui's around. I had a closer look at them to see what is already around and especially 
http://www.swine-tool.de/
looked interesting.
http://linuxpoison.blogspot.com/2009/09/winexs-gui-graphical-user-interface-for.html
this one too.

What do you think on above interfaces? 
Swine is already Qt4, porting to Qt5 should not be hard and it should be platform independent too.

**Comment by slackner** [comment]
Hi everyone,

About one year ago, we announced that we would like to move to more advanced bug tracker. We are sorry for the inconvenience, but now its time to say finally goodbye to the old one. This means, all remaining opened bugs (including this one!) will be closed, and no longer available. If you still need any information from the old bugtracker, please write a mail to webmaster@fds-team.de.

More important, if your issue still exists, please open a new bug report at one of the places mentioned below. For bug reports, please attach terminal output from a recent Wine version (preferably >= 1.8). Please note that in the meantime, Wine Staging has become an official part of WineHQ, so if your issue is already tracked at winehq.org, there is NO NEED to open a new bug report for it. Feel free to add the Wine Staging maintainers (sebastian@fds-team.de, michael@fds-team.de) as CC if you want us to take a look at your bug report.

**NEW BUG TRACKERS:**

* For all Wine AND Wine Staging bugs. USE THIS IF IN DOUBT. (*)
  https://bugs.winehq.org/enter_bug.cgi?product=Wine

* For packaging issues:
  https://bugs.winehq.org/enter_bug.cgi?product=Packaging

* For patch submission:
  https://bugs.wine-staging.com/enter_bug.cgi

Best regards,
Sebastian

----

(*) The "Wine Staging" product in the winehq bugtracker should only be used, when your problem turns out to be a regression introduced in "Wine Staging". Since most problems also occur with Wine, it is recommended to open your bug in the "Wine" product, unless someone can show that it doesn't occur with the development branch.
